pkg/operator/agent/agent.go

```diff
@@ -0,0 +1,250 @@
+		clientset: clientset,
+	}
+}
+
+// Start the agent
+func (a *Agent) Start(namespace string) error {
+	err := a.createAgentDaemonSet(namespace)
+	if err != nil {
+		return fmt.Errorf("Error starting agent daemonset: %v", err)
+	}
+	return nil
+}
+
+func (a *Agent) createAgentDaemonSet(namespace string) error {
+
+	serviceAccount := os.Getenv(k8sutil.RookOperatorServiceAccount)
+
+	agentImage, err := getContainerImage(a.clientset)
+	if err != nil {
+		return err
+	}
+	privileged := true
+	ds := &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: agentDaemonsetName,
+		},
+		Spec: extensions.DaemonSetSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": agentDaemonsetName,
+					},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: serviceAccount,
+					Containers: []v1.Container{
+						{
+							Name:  agentDaemonsetName,
+							Image: agentImage,
+							Args:  []string{"agent"},
+							SecurityContext: &v1.SecurityContext{
+								Privileged: &privileged,
+							},
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "flexvolume",
+									MountPath: "/flexmnt",
+								},
+								{
+									Name:      "dev",
+									MountPath: "/dev",
+								},
+								{
+									Name:      "sys",
+									MountPath: "/sys",
+								},
+								{
+									Name:      "libmodules",
+									MountPath: "/lib/modules",
+								},
+							},
+							Env: []v1.EnvVar{
+								k8sutil.NamespaceEnvVar(),
+								k8sutil.NodeEnvVar(),
+							},
+						},
+					},
+					Volumes: []v1.Volume{
+						{
+							Name: "flexvolume",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: a.discoverFlexvolumeDir(),
+								},
+							},
+						},
+						{
+							Name: "dev",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/dev",
+								},
+							},
+						},
+						{
+							Name: "sys",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/sys",
+								},
+							},
+						},
+						{
+							Name: "libmodules",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/lib/modules",
+								},
+							},
+						},
+					},
+					HostNetwork: true,
+					Tolerations: []v1.Toleration{
+						{
+							Effect:   v1.TaintEffectNoSchedule,
+							Operator: v1.TolerationOpExists,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	_, err = a.clientset.Extensions().DaemonSets(namespace).Create(ds)
+	if err != nil {
+		if !kserrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create rook-agent daemon set. %+v", err)
+		}
+		logger.Infof("rook-agent daemonset already exists")
+	} else {
+		logger.Infof("rook-agent daemonset started")
+	}
+	return nil
+
+}
+
+func (a *Agent) discoverFlexvolumeDir() string {
+	//copy flexvolume to flexvolume dir
```

Comment:
no actual copying is done by this function, right?

Reply:
correct. This is just to discover where the flexvolume driver dir is. This is needed for two things:
1- Copying the rook driver to this folder (done by the agent)
2- To start agent with mount hostpath on this folder.